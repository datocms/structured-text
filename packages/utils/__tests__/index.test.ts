import {
  allowedNodeTypes,
  allowedChildren,
  allowedAttributes,
  isHeading,
  render,
  StructuredText,
  renderRule,
  Node,
  RenderError,
} from '../src';

describe('datocms-structured-text-utils', () => {
  describe('definitions', () => {
    it('are coherent', () => {
      expect(allowedNodeTypes).toEqual(Object.keys(allowedChildren));
      expect(allowedNodeTypes).toEqual(Object.keys(allowedAttributes));
      expect(
        Object.entries(allowedAttributes)
          .filter((entry) => entry[1].includes('children'))
          .map((entry) => entry[0]),
      ).toEqual(
        Object.entries(allowedChildren)
          .filter((entry) => entry[1].length > 0)
          .map((entry) => entry[0]),
      );
      expect(
        Object.entries(allowedAttributes)
          .filter((entry) => !entry[1].includes('children'))
          .map((entry) => entry[0]),
      ).toEqual(
        Object.entries(allowedChildren)
          .filter((entry) => entry[1].length === 0)
          .map((entry) => entry[0]),
      );
    });
  });

  describe('guards', () => {
    it('work as expected', () => {
      expect(isHeading({ type: 'blockquote', children: [] })).toBeFalsy();
      expect(
        isHeading({ type: 'heading', level: 3, children: [] }),
      ).toBeTruthy();
    });
  });

  describe('render', () => {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    const dummyRenderer = (context: any): any => {
      return context;
    };

    const adapter = {
      renderNode: dummyRenderer,
      renderFragment: (chunks: string[]) => chunks,
      renderText: (text: string) => text,
    };

    describe('null value', () => {
      it('returns null', () => {
        expect(render(adapter, null, [])).toMatchInlineSnapshot(`null`);
      });
    });

    describe('some value', () => {
      const structuredText: StructuredText = {
        value: {
          schema: 'dast',
          document: {
            type: 'root',
            children: [
              {
                type: 'paragraph',
                children: [
                  { type: 'span', marks: ['strikethrough'], value: 'Foobar' },
                  { type: 'inlineBlock', item: '123' },
                ],
              },
            ],
          },
        },
        blocks: [],
        links: [],
      };

      describe('no rules', () => {
        it('returns null', () => {
          expect(() => {
            render(adapter, structuredText, []);
          }).toThrow(RenderError);
        });
      });

      describe('some rules', () => {
        it('returns null', () => {
          expect(
            render(adapter, structuredText, [
              renderRule(
                (node: Node): node is Node => true,
                ({ adapter, ...other }) => {
                  return adapter.renderNode(other);
                },
              ),
            ]),
          ).toMatchSnapshot();
        });
      });
    });
  });
});
