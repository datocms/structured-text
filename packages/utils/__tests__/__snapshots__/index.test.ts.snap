// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`datocms-structured-text-utils render some value some rules returns null 1`] = `
Object {
  "ancestors": Array [],
  "children": Array [
    Object {
      "ancestors": Array [
        Object {
          "children": Array [
            Object {
              "children": Array [
                Object {
                  "marks": Array [
                    "strikethrough",
                  ],
                  "type": "span",
                  "value": "Foobar",
                },
                Object {
                  "item": "123",
                  "type": "inlineBlock",
                },
              ],
              "type": "paragraph",
            },
          ],
          "type": "root",
        },
      ],
      "children": Array [
        Object {
          "ancestors": Array [
            Object {
              "children": Array [
                Object {
                  "marks": Array [
                    "strikethrough",
                  ],
                  "type": "span",
                  "value": "Foobar",
                },
                Object {
                  "item": "123",
                  "type": "inlineBlock",
                },
              ],
              "type": "paragraph",
            },
            Object {
              "children": Array [
                Object {
                  "children": Array [
                    Object {
                      "marks": Array [
                        "strikethrough",
                      ],
                      "type": "span",
                      "value": "Foobar",
                    },
                    Object {
                      "item": "123",
                      "type": "inlineBlock",
                    },
                  ],
                  "type": "paragraph",
                },
              ],
              "type": "root",
            },
          ],
          "children": undefined,
          "key": "t-0",
          "node": Object {
            "marks": Array [
              "strikethrough",
            ],
            "type": "span",
            "value": "Foobar",
          },
        },
        Object {
          "ancestors": Array [
            Object {
              "children": Array [
                Object {
                  "marks": Array [
                    "strikethrough",
                  ],
                  "type": "span",
                  "value": "Foobar",
                },
                Object {
                  "item": "123",
                  "type": "inlineBlock",
                },
              ],
              "type": "paragraph",
            },
            Object {
              "children": Array [
                Object {
                  "children": Array [
                    Object {
                      "marks": Array [
                        "strikethrough",
                      ],
                      "type": "span",
                      "value": "Foobar",
                    },
                    Object {
                      "item": "123",
                      "type": "inlineBlock",
                    },
                  ],
                  "type": "paragraph",
                },
              ],
              "type": "root",
            },
          ],
          "children": undefined,
          "key": "t-1",
          "node": Object {
            "item": "123",
            "type": "inlineBlock",
          },
        },
      ],
      "key": "t-0",
      "node": Object {
        "children": Array [
          Object {
            "marks": Array [
              "strikethrough",
            ],
            "type": "span",
            "value": "Foobar",
          },
          Object {
            "item": "123",
            "type": "inlineBlock",
          },
        ],
        "type": "paragraph",
      },
    },
  ],
  "key": "t-0",
  "node": Object {
    "children": Array [
      Object {
        "children": Array [
          Object {
            "marks": Array [
              "strikethrough",
            ],
            "type": "span",
            "value": "Foobar",
          },
          Object {
            "item": "123",
            "type": "inlineBlock",
          },
        ],
        "type": "paragraph",
      },
    ],
    "type": "root",
  },
}
`;
